client/commands/v2/pysa_server.py

```diff
@@ -59,6 +65,61 @@ def __init__(
         self.pyre_arguments = pyre_arguments
         self.binary_location = binary_location
         self.server_identifier = server_identifier
+        self.pyre_connection = api_connection.PyreConnection(
+            Path(self.pyre_arguments.global_root)
+        )
+        self.file_tracker: Set[Path] = set()
+
+    def invalid_model_to_diagnostic(
+        self, invalid_model: query.InvalidModel
+    ) -> lsp.Diagnostic:
+        return lsp.Diagnostic(
+            range=lsp.Range(
+                start=lsp.Position(
+                    line=invalid_model.line - 1, character=invalid_model.column
+                ),
+                end=lsp.Position(
+                    line=invalid_model.stop_line - 1,
+                    character=invalid_model.stop_column,
+                ),
+            ),
+            message=invalid_model.full_error_message,
+            severity=lsp.DiagnosticSeverity.ERROR,
+            code=None,
+            source="Pysa",
+        )
+
+    def invalid_models_to_diagnostics(
+        self, invalid_models: Sequence[query.InvalidModel]
+    ) -> Dict[Path, List[lsp.Diagnostic]]:
+        result: Dict[Path, List[lsp.Diagnostic]] = defaultdict(list)
+        for model in invalid_models:
+            if model.path is None:
+                self.log_and_show_message_to_client(
+                    f"{model.full_error_message}", lsp.MessageType.ERROR
+                )
+            else:
+                result[Path(model.path)].append(self.invalid_model_to_diagnostic(model))
+        return result
+
+    async def update_errors(self) -> None:
+        # Publishing empty diagnostics to clear errors in VSCode and reset self.file_tracker
+        for document_path in self.file_tracker:
+            await _publish_diagnostics(self.output_channel, document_path, [])
+        self.file_tracker = set()
+
+        try:
+            model_errors = query.get_invalid_taint_models(self.pyre_connection)
+            diagnostics = self.invalid_models_to_diagnostics(model_errors)
+            # Keep track of files we publish diagnostics for
+            for path in diagnostics.keys():
```

Comment:
No need to loop. `set.update` lets you provide a iterable to add into the set
```suggestion
            self.file_tracker.update(diagnostics.keys())
```